src/coreclr/pal/tests/palsuite/threading/SleepEx/test2/test2.cpp

```diff
@@ -160,6 +168,7 @@ DWORD PALAPI SleeperProc_SleepEx_test2(LPVOID lpParameter)
     }
 
     OldTimeStamp = GetHighPrecisionTimeStamp(performanceFrequency);
+    s_preWaitTimestampRecorded = true;
```

Comment:
The variable being volatile is not sufficient on architectures like ARM. We need to add memory barriers at read and write locations or use interlocked operations that have the barriers implicit.

Reply:
My figuring was that the read around the Sleep would not need any barriers or volatility because the Sleep would have to imply a barrier if necessary, and for the write the only reason for volatile is to make sure it happens after the clock read, which I don't think is actually necessary, but doesn't hurt at all in this case, and the write is immediately followed by a wait that would imply a barrier where necessary. Even in the worst-case situation, it seems like the order of operations demanded by the code would be ensured by cache consistency at least. Is there a theoretical case where the intended behavior would not work on ARM?

Reply:
I have no problem with using interlocked operations for all of the new memory operations. It's only a test after all, but I'm just curious for my own understanding.

Reply:
I think you are right, I have not realized that the Sleep would induce a memory barrier. My point was that the other thread might never observe the change in the variable value without a barrier.

Reply:
> My point was that the other thread might never observe the change in the variable value without a barrier.

The wait following the write guarantees that it would be visible to other threads, and the Sleep around the read would guarantee that the read would in reasonable time see the update. It would depend on the implementation of Sleep though, a context switch would have to imply a barrier, a Sleep that does not switch would not have to and that could be a problem, but generally the pattern is to imply a barrier on any sort of wait even if it doesn't yield. For instance I'm pretty sure the PAUSE instruction implies a full barrier just because of the delay it implies. If that reasoning is not sufficient though it could easily be changed to a Sleep(1).

Reply:
> If that reasoning is not sufficient

Actually in that case the volatile takes care of it, as it would rely on cache consistency. There are many places in the runtime where we rely entirely on cache consistency for correctness. There are some old processors that don't have any cache consistency and rely entirely on software, but I believe all of the modern processors have sufficient cache consistency behavior to meet our expectations.